### PR TITLE
Enable conversion of negative radian angle values

### PR DESCRIPTION
**Pull request by martiniil**

### CODE REVIEW

**Comment by agutenkunst** [comment]
I am not sure what this does. Could you provide some description in the PR? Sounds like negative rad end up being converted to positive tenthDegree...

**Review by agutenkunst — CHANGES_REQUESTED** [review]
Please explain reasoning

**Comment by agutenkunst** [comment]
> * The class `TenthOfDegree` represents degree values in the range [0, 360] multiplied by ten.
> * The static method `TenthOfDegree::fromRad()` can also convert negative radian values to a `TenthOfDegree` object.

I think the confusing thing would be that `TenthOfDegree::fromRad(-1).toRad() != -1` maybe having a signed and an unsigned version ob this could help?

**Comment by agutenkunst** [comment]
Or we use `int32_t` internally and provide checks where it MUST be positive (only StartRequest comes to my mind...)

**Comment by martiniil** [comment]
> Or we use `int32_t` internally and provide checks where it MUST be positive (only StartRequest comes to my mind...)

Would it be acceptable to cast a positive `int32_t` to `uint16_t`?

**Comment by agutenkunst** [comment]
> > Or we use `int32_t` internally and provide checks where it MUST be positive (only StartRequest comes to my mind...)
> 
> Would it be acceptable to cast a positive `int32_t` to `uint16_t`?

Only if it fits...

File: include/psen_scan_v2/angle_conversions.h [hunk 1]
```diff
@@ -30,29 +33,42 @@ inline static constexpr double degreeToRadian(const double& angle_in_degree)
...
   return degreeToTenthDegree(radianToDegree(angle_in_rad));
 }
 
-inline static double constexpr tenthDegreeToRad(const uint16_t& angle_in_tenth_degree)
+inline static double constexpr tenthDegreeToRad(const int32_t& angle_in_tenth_degree)
 {
   return degreeToRadian(static_cast<double>(angle_in_tenth_degree) / 10.);
 }
 
+inline static uint16_t tenthDegreeToPositiveTenthDegree(const int32_t& angle_in_tenth_degree)
```
Comment (SansoneG) [inline include/psen_scan_v2/angle_conversions.h:59]: I feel like `tenthDegreeToPositiveTenthDegree` and `degreeToTenthDegree` should return the same type. Any reason against this?

File: include/psen_scan_v2/tenth_of_degree.h [hunk 2]
```diff
@@ -49,7 +54,7 @@ class TenthOfDegree
     return *this;
   }
 
-  constexpr TenthOfDegree& operator*(const unsigned int& rhs)
+  constexpr TenthOfDegree& operator*(const int& rhs)
```
Comment (SansoneG) [inline include/psen_scan_v2/tenth_of_degree.h:57]: What happens then this overflows?

File: include/psen_scan_v2/tenth_of_degree.h [hunk 3]
```diff
@@ -28,12 +28,17 @@ class TenthOfDegree
...
+  }
+
 public:
-  explicit constexpr TenthOfDegree(const uint16_t& tenth_of_degree) : tenth_of_degree_(tenth_of_degree)
+  explicit constexpr TenthOfDegree(const int32_t& tenth_of_degree) : tenth_of_degree_(tenth_of_degree)
   {
   }
 
-  constexpr uint16_t value() const
+  constexpr int32_t value() const
```
Comment (SansoneG) [inline include/psen_scan_v2/tenth_of_degree.h:41]: I feel like `value()` and `toPositiveValue` should return the same type

File: test/unit_tests/unittest_start_request.cpp [hunk 4]
```diff
@@ -98,9 +98,11 @@ TEST_F(StartRequestTest, constructorTest)
...
   EXPECT_TRUE(DecodingEquals<uint8_t>(data, static_cast<size_t>(Offset::DIAGNOSTICS_ENABLED), 0));
 
-  EXPECT_TRUE(DecodingEquals(data, static_cast<size_t>(Offset::MASTER_START_ANGLE), scan_range.getStart().value()));
-  EXPECT_TRUE(DecodingEquals(data, static_cast<size_t>(Offset::MASTER_END_ANGLE), scan_range.getEnd().value()));
-  EXPECT_TRUE(DecodingEquals(data, static_cast<size_t>(Offset::MASTER_ANGLE_RESOLUTION), degreeToTenthDegree(0.2)));
+  EXPECT_TRUE(
+      DecodingEquals(data, static_cast<size_t>(Offset::MASTER_START_ANGLE), (uint16_t)scan_range.getStart().value()));
+  EXPECT_TRUE(
+      DecodingEquals(data, static_cast<size_t>(Offset::MASTER_END_ANGLE), (uint16_t)scan_range.getEnd().value()));
+  EXPECT_TRUE(DecodingEquals(data, static_cast<size_t>(Offset::MASTER_ANGLE_RESOLUTION), (uint16_t)2));
```
Comment (SansoneG) [inline test/unit_tests/unittest_start_request.cpp:105]:
```suggestion
      DecodingEquals(data, static_cast<size_t>(Offset::MASTER_START_ANGLE), static_cast<uint16_t>(scan_range.getStart().value())));
  EXPECT_TRUE(
      DecodingEquals(data, static_cast<size_t>(Offset::MASTER_END_ANGLE), static_cast<uint16_t>(scan_range.getEnd().value())));
  EXPECT_TRUE(DecodingEquals(data, static_cast<size_t>(Offset::MASTER_ANGLE_RESOLUTION), static_cast<uint16_t>(2)));
```

File: include/psen_scan_v2/angle_conversions.h [hunk 5]
```diff
@@ -30,29 +33,42 @@ inline static constexpr double degreeToRadian(const double& angle_in_degree)
...
+{
+  if (angle_in_tenth_degree < std::numeric_limits<uint16_t>::min() ||
```
Comment (SansoneG) [inline include/psen_scan_v2/angle_conversions.h:61]: Why does a function that has `toPositive` in it reject negative numbers? This function basically only checks whether a given angle is positive and could therefore also return a `bool` at that point.

File: include/psen_scan_v2/angle_conversions.h (hunk 1 above)
Comment (martiniil) [reply include/psen_scan_v2/angle_conversions.h]: Actually the method name was misleading. I renamed it to `tenthDegreeToUnsignedTenthDegree()`.

File: include/psen_scan_v2/tenth_of_degree.h (hunk 3 above)
Comment (martiniil) [reply include/psen_scan_v2/tenth_of_degree.h:41]: I renamed `toPositiveValue` to `toUnsigned`. I hope now the intention is clear.

File: include/psen_scan_v2/angle_conversions.h (hunk 5 above)
Comment (martiniil) [reply include/psen_scan_v2/angle_conversions.h:61]: Again, the method-name was misleading. I hope now the intention is clear.

File: include/psen_scan_v2/tenth_of_degree.h (hunk 2 above)
Comment (martiniil) [reply include/psen_scan_v2/tenth_of_degree.h:57]: I don't know. Do you suggest that we detect overflow of integer arithmetic?

**Comment by martiniil** [comment]
Needs more revision. Close in favor of #84 for the sake of clarity.

**Pull request closed without merging**